### PR TITLE
Add PC_RotateQuaternion

### PR DESCRIPTION
**Pull request by elemoine**
This PR adds `PC_RotateQuaternion` to rotate patches based on rotation quaternions.

### CODE REVIEW

**Comment by elemoine** [comment]
This is WIP. DO NOT MERGE.

File: lib/pc_patch.c
Comment (mbredif) [inline lib/pc_patch.c]: for performance, you should do the dimx, dimy, dimz lookups only once per patch

File: lib/pc_patch.c
Comment (mbredif) [inline lib/pc_patch.c]: should be freed after the loop

File: lib/pc_patch.c
Comment (mbredif) [inline lib/pc_patch.c]: can't we modify patch in place ? here, values of other dimensions are left unassigned

File: lib/pc_patch.c
Comment (mbredif) [inline lib/pc_patch.c]: for a linear rotation, a PCMAT33 and a VEC3 could be used instead

Comment (elemoine) [reply lib/pc_patch.c]: Right, will change this.

File: lib/pc_matrix.c
Comment (mbredif) [inline lib/pc_matrix.c]: why do you use a PCMAT44 for a normalized quaternion ? I suggest to use a linear PCMAT33 for a normalized quaternion and a projective PCMAT44 for an unnormalized quaternion, replacing 1's by |q|^2 on the diagonal.

File: lib/pc_patch.c
Comment (elemoine) [reply lib/pc_patch.c]: `PC_RotateQuaternion` returns a new `PcPatch` so I don't think an in-place modification is possible. Regarding, the other dimensions, I think they should be copied in the new `PcPatch`.

File: lib/pc_patch.c
Comment (elemoine) [reply lib/pc_patch.c]: You mean do a lookup for the very first point and then use indices?

File: lib/pc_patch.c
Comment (elemoine) [reply lib/pc_patch.c]: Right, will change.

File: lib/pc_matrix.c
Comment (elemoine) [reply lib/pc_matrix.c]: As already said, I'll use `PCMAT33` and `PCVEC3` only for this first patch.

File: lib/pc_patch.c
Comment (mbredif) [reply lib/pc_patch.c]: I think the pointlist is a temporary copy already so that it can be modified in place, but that should be checked obviously.

Comment (elemoine) [reply lib/pc_patch.c]: Yep, you may be right. I'll check.

File: pgsql/pointcloud.sql.in
Comment (elemoine) [inline pgsql/pointcloud.sql.in]: Missing space before `AS` here.

File: lib/pc_patch.c
Comment (elemoine) [inline lib/pc_patch.c]: To be removed when a 3x3 matrix will be used.

File: lib/pc_patch.c
Comment (mbredif) [reply lib/pc_patch.c]: Yes, I mean `dim = pc_schema_get_dimension_by_name(schema, name);` once then do `pc_point_get_double(pt, dim, d);` for each point.

**Comment by elemoine** [comment]
@mbredif I addressed your (good) comments. I need to write more tests.

**Comment by elemoine** [comment]
@mbredif feel free to review this again.

**Comment by mbredif** [comment]
Looks good !
3 comments, addressing the first one being sufficient for now in my opinion : 
- A comment is needed for future maintainers of the patch version of PC_RotateQuaternion, explaining that points iterated from pointlists extracted from uncompressed patches are just pointers to the raw patch buffer, so that updating the points directly changes the patch payload.
- x and y dimensions should ideally default to the `x_position` and `y_position`. `z_position` is however not provided. I propose to defer that to another issue (https://github.com/LI3DS/pointcloud/issues/32)
- point rotation through quaternions can be optimized in terms of floating point operations and memory allocation (see : [Quaternions_and_spatial_rotation](https://en.wikipedia.org/w/index.php?title=Quaternions_and_spatial_rotation&section=26#Comparison_with_other_representations_of_rotations) / Used methods 2)

**Comment by elemoine** [comment]
> A comment is needed for future maintainers of the patch version of PC_RotateQuaternion, explaining that points iterated from pointlists extracted from uncompressed patches are just pointers to the raw patch buffer, so that updating the points directly changes the patch payload.

I added a comment.

> x and y dimensions should ideally default to the x_position and y_position. z_position is however not provided. I propose to defer that to another issue

Agree.

**Comment by elemoine** [comment]
> point rotation through quaternions can be optimized in terms of floating point operations and memory allocation (see : Quaternions_and_spatial_rotation / Used methods 2)

Please be more specific. I don't know undertand what method you're referring to.

**Comment by elemoine** [comment]
I think I found it. I'll give it a try.

**Comment by mbredif** [comment]
The most efficient way to rotate a vector with a quaternion is {\vec {v}}_{\text{new}}={\vec {v}}+2{\vec {r}}\times ({\vec {r}}\times {\vec {v}}+w{\vec {v}}) ([wikipedia](https://en.wikipedia.org/w/index.php?title=Quaternions_and_spatial_rotation&section=26#Performance_comparisons)), which needs 15 multiplications and 15 additions to evaluate (or 18 muls and 12 adds if the factor of 2 is done via multiplication.), compared to the optimal 21 multiplications ( = 12 + 9 ) and 18 add/subtraction ( = 12 + 6 ) with the method of this PR (matrix creation + matrix application).

Considering that the proposed matrix creation method `pc_matrix_create_from_quaternion` is currently not optimized (6 pow, 30 multiplications, 12 add/sub rather than the optimal 0 pow,12 mult,12 add/sub ), I suggest to accept this PR as is and open an issue for this optimization if that becomes a bottleneck, ok?

**Comment by elemoine** [comment]
I am currently trying to make Method 2 work, but it does not – my unit tests fail. I don't understand why.

**Comment by elemoine** [comment]
@mbredif please take a look at https://github.com/elemoine/pointcloud/commit/0b6b2e558688d945d481fceb7bfbdac1a31f869d if you have some time. I cannot find what it is not correct and why it doesn't work.

**Comment by elemoine** [comment]
http://gamedev.stackexchange.com/questions/28395/rotating-vector3-by-a-quaternion is also relevant.

**Comment by mbredif** [comment]
Thinking back, actually, you construct the matrix once and apply it for each point so its construction cost is amortized and your solution is indeed optimal as soon as you have more than a few points.  The direct quaternion-based rotation ( elemoine@0b6b2e5 ) will thus only be relevant for version where the quaternion changes for each point... Sorry for the confusion.

**Comment by elemoine** [comment]
> Thinking back, actually, you construct the matrix once and apply it for each point so its construction cost is amortized and your solution is indeed optimal as soon as you have more than a few points. The direct quaternion-based rotation ( elemoine@0b6b2e5 ) will thus only be relevant for version where the quaternion changes for each point... Sorry for the confusion.

Good point. That being said I still don't understand why https://github.com/elemoine/pointcloud/commit/0b6b2e5 doesn't work.

**Comment by elemoine** [comment]
Before we merge this I'd like to do another change. I think the `pc_patch|point_rotate_quaternion` functions could place `vec` and `rvec` on the stack instead of using `malloc`.

**Comment by elemoine** [comment]
> Before we merge this I'd like to do another change. I think the pc_patch|point_rotate_quaternion functions could place vec and rvec on the stack instead of using malloc.

Done.

**Comment by mbredif** [comment]
Could you review mbredif/pointcloud/rotate that enforce const correctness (point rotation is performed in place, which is kind of weird compared to patch rotation which is returning a new patch) and computes rotation matrix on the stack.

**Comment by elemoine** [comment]
@mbredif https://github.com/mbredif/pointcloud/commit/60cd5c94cd14cc08420e399d7587adb4ba5361c9 in your branch includes many unrelated things.

**Comment by mbredif** [comment]
My mistake, I cleaned that up in mbredif@a92c265.

**Comment by elemoine** [comment]
@mbredif I updated the patch based on your commits. This looks good. Thanks for your great suggestions.

**Review by mbredif — CHANGES_REQUESTED** [review]
Looks good, just a little concern in the comment above.

**Comment by elemoine** [comment]
> Looks good, just a little concern in the comment above.

What comment?

File: pgsql/pc_editor.c
Comment (mbredif) [inline pgsql/pc_editor.c:63]: only required if patch_in has been cloned or decompressed rather than modified in place

File: lib/pc_patch.c
Comment (mbredif) [inline lib/pc_patch.c:684]: My understanding is that you need to pc_patch_uncompressed_clone the patch to ensure that you can modify it.
- going back and forth to a pointlist does not make that logic apparent
- do you really need to clone if the patch is not readonly ? (is the patch readonly in this PR's use case ?)

Comment (elemoine) [reply lib/pc_patch.c:684]: In the case where the input patch is not compressed we could do the rotation in place. But for a consistent API, where the returned patch is alway a new patch, I decided to clone the input patch and do the rotation in that cloned patch.

File: pgsql/pc_editor.c
Comment (elemoine) [reply pgsql/pc_editor.c:63]: As I just commented the output patch is always a new patch. So the user of `pc_patch_rotate_quaternion` is responsible for freeing that output patch.

File: lib/pc_patch.c
Comment (mbredif) [reply lib/pc_patch.c:684]: Concerning the api, that pattern is already used in the lib, for instance :
- [Conditionnal clone](https://github.com/pgpointcloud/pointcloud/blob/3929653f51296f5dd5fe5997c8c4c5d46419cb50/pgsql/pc_access.c#L579)
- [Conditionnal free](https://github.com/pgpointcloud/pointcloud/blob/3929653f51296f5dd5fe5997c8c4c5d46419cb50/pgsql/pc_access.c#L622)

a `pc_patch_uncompressed_clone` function (and its siblings) would enable more readable code and optimized execution.

Comment (elemoine) [reply lib/pc_patch.c:684]: Agree with `pc_patch_uncompressed_clone`, but I did not want to put too much in that patch.

Comment (elemoine) [reply lib/pc_patch.c:684]: > Concerning the api, that pattern is already used in the lib, for instance

So do you mean that my implementation in fine w.r.t. this?

**Comment by elemoine** [comment]
@mbredif more comments on this PR? Should we merge it?

**Review by mbredif — APPROVED** [review]

**Comment by mbredif** [comment]
Thanks @elemoine for this PR.
I deferred my little concerns to #38, #39.

**Pull request merged**
